log4j-core/src/main/java/org/apache/logging/log4j/core/impl/CoreDefaultBundle.java

```diff
@@ -103,24 +105,58 @@ public FlowMessageFactory defaultFlowMessageFactory() {
 
     @SingletonFactory
     @ConditionalOnMissingBinding
-    public RecyclerFactoryProvider defaultRecyclerFactoryProvider(
+    public RecyclerFactory defaultRecyclerFactory(
             final PropertyEnvironment environment,
             final ClassLoader loader,
             final @Named("StatusLogger") org.apache.logging.log4j.Logger statusLogger) {
-        final String factory = environment.getProperty(RecyclerProperties.class).factory();
-        final Stream<RecyclerFactoryProvider> providerStream = ServiceLoaderUtil.safeStream(
-                RecyclerFactoryProvider.class, ServiceLoader.load(RecyclerFactoryProvider.class, loader), statusLogger);
-        final Optional<RecyclerFactoryProvider> provider = factory != null
-                ? providerStream.filter(p -> factory.equals(p.getName())).findAny()
-                : providerStream.min(Comparator.comparing(RecyclerFactoryProvider::getOrder));
-        return provider.orElseGet(RecyclerFactoryProvider::getInstance);
-    }
 
-    @SingletonFactory
-    @ConditionalOnMissingBinding
-    public RecyclerFactory defaultRecyclerFactory(
-            final PropertyEnvironment environment, final RecyclerFactoryProvider provider) {
-        return provider.createForEnvironment(environment);
+        // Collect providers
+        @Nullable
+        final String providerName =
+                environment.getProperty(RecyclerProperties.class).factory();
+        final List<RecyclerFactoryProvider> providers = ServiceLoaderUtil.safeStream(
+                        RecyclerFactoryProvider.class,
+                        ServiceLoader.load(RecyclerFactoryProvider.class, loader),
+                        statusLogger)
+                .sorted(Comparator.comparing(RecyclerFactoryProvider::getOrder))
```

Comment:
Can we use a composite comparator such that:

1. The provider supplied using `log4j.recycler.factory` comes first,
2. The remaining providers are ordered by `Order`.

```java
.sorted(Comparator.comparing(p -> p.getName().equals(providerName) ? Integer.MIN_VALUE : p.getOrder()
```

should work

Reply:
I prefer a more dumb-proof style: Have an explicit block for `if (providerName != null)`.